## Analysis of used cars from eBay Kleinanzeigen
This is a German eBay website. The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). We've made a few modifications from the original dataset that was uploaded to Kaggle.


The aim of this project is:
* clean the data
* analyze the included used car listings

In [1]:
import pandas as pd
import numpy as np

In [2]:
# if read in file without assinging encoding
# it will default to "UTF-8"
# try "Latin-1" or "Windows-1252" until you're able to read the file
autos = pd.read_csv("autos.csv", encoding = "Windows-1252")

In [3]:
# run the variable "autos" to review data
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
# see the information of autos
print(autos.describe(include = "all"))

                dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44

We can make the following observations:
* The dataset contains 20 columns, most of which are strings.
* Some columns have null values, but none have more than ~20% null values.
* The column name use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscore.

In [5]:
autos = autos.drop(["nrOfPictures","seller", "offerType"], axis = 1)

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'price', 'abtest', 'vehicleType',
       'yearOfRegistration', 'gearbox', 'powerPS', 'model', 'odometer',
       'monthOfRegistration', 'fuelType', 'brand', 'notRepairedDamage',
       'dateCreated', 'postalCode', 'lastSeen'],
      dtype='object')

In [7]:
# rename below columns with funtion "clean_column"
def cleaning(col):
    col = col.replace("yearOfRegistration", "registration_year")
    col = col.replace("monthOfRegistration", "registration_month")
    col = col.replace("notRepairedDamage", "unrepaired_damage")
    col = col.replace("dateCreated", "ad_created")
    col = col.lower()
    return col

# use an empty list and a for-loop to rename column names
# with funtion "cleaning"
cleaned_column = []
for col in autos.columns:
    updated_column = cleaning(col)
    cleaned_column.append(updated_column)
    
# print(cleaned_column) to check
# assign the modified column names back to DataFrame.columns attribute
autos.columns = cleaned_column


In [8]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicletype  registration_year    gearbox  powerps   model  \
0  $5,000  control         bus               2004    manuell      158  andere   
1  $8,500  control   limousine               1997  automatik      286     7er   
2  $8,990     test   limousine               2009    manuell      102    golf   
3  $4,350  control  kleinwagen               2007  automatik       71  fortwo   
4  $1,350     test       kombi               2003    manuell        0   focus   

    odometer  registration_month fueltype       brand unrepaired_damage  \
0  150,000km                   3      lpg     peugeot              nein   
1  150,000km                   6   benzin         bmw              nein   
2   70,000km                   7   benzin  volkswagen              nein   
3   70,000km                   6   benzin       smart              nein   
4  150,000km                   7   benzin        ford              nein   

            ad_created  postalcode             lastseen  
0  2016-03-26 00:00:00       79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00       71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00       35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00       33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00       39218  2016-04-01 14:38:50

First, original column names are made of camelcase which is hard to read and select when coding. We define a function "cleaning" to use underscores to seperate two words, to use "DataFrame.lower()" to make column-name snakecase.
Most important of all, remember to return value to avoid None values.

Second, use an empty list and a for-loop to change column names and assign them back.


Finally, print head columns of autos DataFrame to re-check if everything is correct.

In [9]:
# use DataFrame.describe() to look at descriptive statistics
autos.describe()

registration_year       powerps  registration_month    postalcode
count       50000.000000  50000.000000        50000.000000  50000.000000
mean         2005.073280    116.355920            5.723360  50813.627300
std           105.712813    209.216627            3.711984  25779.747957
min          1000.000000      0.000000            0.000000   1067.000000
25%          1999.000000     70.000000            3.000000  30451.000000
50%          2003.000000    105.000000            6.000000  49577.000000
75%          2008.000000    150.000000            9.000000  71540.000000
max          9999.000000  17700.000000           12.000000  99998.000000

* Any columns that have mostly one value that are candidates to be dropped.
* Any columns that need more investigation
* Any examples of numeric data stored as text that needs to be cleaned.

In [10]:
print(autos["price"].describe())
print(autos["odometer"].describe())

count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object
count         50000
unique           13
top       150,000km
freq          32424
Name: odometer, dtype: object


In [11]:
# we found "price" & "odometer" columns are
# numeric values stored as text
# use Series.str.replace("original","updated")
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

# use DataFrame.rename() to rename column to "odometer_km"
autos.rename({"odometer":"odometer_km"}, axis = 1, inplace = True)

In [12]:
# see how many unique values of price
autos["price"].unique().shape

(2357,)

In [13]:
autos["price"].sort_values(ascending = True).tail(20)
# you can see price suddenly rise from 350,000 to $999,990
# we'd like to remove outliers since 350,000

47337      259000
12682      265000
35923      295000
34723      299000
14715      345000
36818      350000
37585      999990
514        999999
43049      999999
22947     1234566
7814      1300000
47634     3890000
11137    10000000
24384    11111111
2897     11111111
27371    12345678
47598    12345678
39377    12345678
42221    27322222
39705    99999999
Name: price, dtype: int64

In [14]:
autos["odometer_km"].value_counts(ascending = False)
# no obvious data need to be removed

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [15]:
autos = autos[autos["price"].between(1,350001)]
# bol_1 = autos["price"] < 350000
# bol_2 = autos["price"] > 0
# autos["price"] = autos[bol_1 & bol_2]

In [16]:
# 1,435 outlier data was removed
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [17]:
# "series.value_counts()" to see distribution
# "normalize = True" to see percentage
# chain "series.sort_values(ascending = True)" to see earliest to latest
# date_crawled
    
date_crawled = autos["datecrawled"].str[:10]
date_crawled = date_crawled.value_counts(normalize = True, dropna = False).sort_index(ascending = True)
print(date_crawled.head())


2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
Name: datecrawled, dtype: float64


In [18]:
# ad_created

ad_created = autos["ad_created"].str[:10]
ad_created = ad_created.value_counts(normalize = True, dropna= False).sort_values(ascending = True)
print(ad_created.head())


2016-01-22    0.000021
2015-09-09    0.000021
2016-02-09    0.000021
2016-01-13    0.000021
2015-12-05    0.000021
Name: ad_created, dtype: float64


In [19]:
# last_seen

last_seen = autos["lastseen"].str[:10]
last_seen = last_seen.value_counts(normalize = True, dropna = False).sort_values(ascending = True)
print(last_seen.head())


2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-18    0.007351
2016-03-08    0.007413
Name: lastseen, dtype: float64


In [20]:
# use Series.describe() to understand the distribution
# of registration_year

autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Based on exploration above, 
* The min value is 1000, before car is invented
* The max value is 9999, many years into the future

According to our common sense, the car was invented in the early 1900s. Besides, the car can't be registered after this data was collected in 2016. 

We'd like to count the number with cars that fall outside the 1900 - 2016. 

Let's research a little bit further.

In [21]:
autos["registration_year"][autos["registration_year"] > 2016].sort_values()
# there's 1879 data which registration year is over 2016

10       2017
30547    2017
30505    2017
30502    2017
30470    2017
30336    2017
30266    2017
30263    2017
30262    2017
30219    2017
30175    2017
30149    2017
30122    2017
30085    2017
30038    2017
30012    2017
29983    2017
29976    2017
29945    2017
29929    2017
29917    2017
29882    2017
30570    2017
29798    2017
30592    2017
30830    2017
31397    2017
31389    2017
31371    2017
31341    2017
         ... 
11364    2018
11264    2018
11360    2018
34111    2018
11355    2018
11436    2018
33851    2018
11722    2018
11774    2018
11290    2018
34107    2018
34099    2018
33738    2018
49185    2019
5763     2019
27578    2800
4549     4100
453      4500
42079    4800
49153    5000
22799    5000
24519    5000
4164     5000
27618    5911
8360     6200
25003    8888
49910    9000
8012     9999
33950    9999
38076    9999
Name: registration_year, Length: 1879, dtype: int64

In [22]:
autos["registration_year"][autos["registration_year"] < 1900].sort_values()
# there's only 5 data which registration year is below 1900
# It's safe to remove outliers since there's only 3.7% data removed

22316    1000
49283    1001
24511    1111
10556    1800
32585    1800
Name: registration_year, dtype: int64

In [23]:
# use boolean to remove value which 
# registration_year is <1900 and >2016
bool_1 = (autos["registration_year"] > 1900)
bool_2 = (autos["registration_year"] < 2017)
autos = autos.loc[bool_1 & bool_2]

# print(autos["registration_year"].describe())
autos["registration_year"].value_counts(normalize = True, ascending = False).head(10)


2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

We found out most registration year was: 2000, 2005, 1999, 2004, 2003,...,etc, which means most used cars are over 10 years and if you purchase them, you might need to take good care to save repair expenses.

In [24]:
print(autos.columns)

Index(['datecrawled', 'name', 'price', 'abtest', 'vehicletype',
       'registration_year', 'gearbox', 'powerps', 'model', 'odometer_km',
       'registration_month', 'fueltype', 'brand', 'unrepaired_damage',
       'ad_created', 'postalcode', 'lastseen'],
      dtype='object')


In [25]:
# identify the unique values we want to aggregate by
brands = autos["brand"].value_counts(normalize=True)
print(brands.head(10))

# select Volkswagen, BMW, Opel, Mercedes Benz, Audi and Ford 
# to analyze its price mean.

volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
Name: brand, dtype: float64


Based on result above, we can tell German really loves German-brand as headquarter of top 5 brands on the list are located in German.
Volkswagen(21%), BMW(11%), Opel(10%), Mercedes Benz(9%) and Audi(8%) are 59% in total.

Except for German brands, no.6 on the list is American-brand Ford(6.9%). no.7 Renault(4.7%) and no.8 Peugeot(2.9%) are French-brand. 

In [26]:
# "brands" is a series with index labels
# use series.index to access the labels
brands_over_5 = brands.index[:6]

In [27]:
# Try to calculate mean price of Volkswagen
# and then finish a for-loop
volkswagen = (autos["brand"] == "volkswagen")
volkswagen_price_mean = autos["price"][volkswagen].mean()
# volkswagen_length = autos["price"][volkswagen].value_counts()
print(volkswagen_price_mean)

# try to use a for-loop to finish mean values of 6 brands

5402.410261610221


In [28]:
# create an empty list to store mean prices
brand_mean_price = {}

# use a for-loop to collect mean price
for brand_name in brands_over_5:
    bool_1 = (autos["brand"] == brand_name)
    mean_price = autos["price"][bool_1].mean().round()
    brand_mean_price[brand_name] = mean_price

print(brand_mean_price)

{'audi': 9337.0, 'volkswagen': 5402.0, 'opel': 2975.0, 'ford': 3749.0, 'mercedes_benz': 8628.0, 'bmw': 8333.0}


Among those brands with over 5% market share in eBay, we can classify brands by price as below.
* Very expensive(>8000): Audi,Mercedes Benz, BMW
* In between: Volkswagen
* Less expensive(<4000): Ford, Opel

Let's use aggregation to understand if there's any link between average mileage and average price.

In [29]:
#  mean mileage of 6 brands
brand_mean_mileage = {}

for brand_mileage in brands_over_5:
    bool_2 = (autos["brand"] == brand_mileage)
    mean_mileage = autos["odometer_km"][bool_2].mean().round()
    brand_mean_mileage[brand_mileage] = mean_mileage
    
print(brand_mean_mileage)


{'audi': 129157.0, 'volkswagen': 128707.0, 'opel': 129310.0, 'ford': 124266.0, 'mercedes_benz': 130788.0, 'bmw': 132573.0}


In [30]:
# convert both dictionaries to series object
bmp_series = pd.Series(brand_mean_price)
bmm_series = pd.Series(brand_mean_mileage)

# create a dataframe from the first series object
# df = pd.DataFrame({"mean_price":bmp_series, "mean_mileage":bmm_series})
df = pd.DataFrame(bmp_series, columns=["mean_price"])

# assign the other series as a new column
df["mean_mileage"] = bmm_series

# see the relationship between price & mileage
print(df.sort_values("mean_price", ascending = False))
print()
print(df.corr())


               mean_price  mean_mileage
audi               9337.0      129157.0
mercedes_benz      8628.0      130788.0
bmw                8333.0      132573.0
volkswagen         5402.0      128707.0
ford               3749.0      124266.0
opel               2975.0      129310.0

              mean_price  mean_mileage
mean_price      1.000000      0.612343
mean_mileage    0.612343      1.000000


For very expensive brands(Audi,Mercedes Benz, BMW), the more mileage, the price is less expensive.

For less expensive brands(Ford, Opel), it's the same that the more mileage, cars would be sold at lower price.

# 接下來的任務
#### 將categorical data從德文翻譯回英文 (finished)
#### 將dates改為數字格式20160321 (finished)
#### 檢查是否有特殊的關鍵字在name欄中，抽取到新的columns

# 接下來的分析
#### 找到最多的品牌/model組合 ex:Volkswagen/Golf (finished)
#### 拆分odometer_km到群組，利用總和檢視平均價格是否與里程數規則相同
#### 車損是否影響價格？會影響多少？

In [38]:
# Below columns have german words and we'd like to translate into English
print(autos["vehicletype"].unique())
print()
print(autos["gearbox"].unique())
print()
print(autos["fueltype"].unique())
print()
print(autos["unrepaired_damage"].unique())

['bus' 'limousine' 'small car' 'combi' nan 'coupe' 'suv' 'convertible'
 'other']

['manually' 'automatic' nan]

['lpg' 'petrol' 'diesel' nan 'cng' 'hybrid' 'electro' 'other']

['nein' nan 'ja']


In [32]:
# build a vehicle type dictionary
vehicletype_dict = {"bus":"bus", "limousine":"limousine",
                    "coupe":"coupe", "suv":"suv",
                    "kleinwagen":"small car", "kombi":"combi",
                   "cabrio":"convertible", "andere":"other"}

# mapping dictionary to column
autos["vehicletype"] = autos["vehicletype"].map(vehicletype_dict)

print(autos["vehicletype"].unique())

['bus' 'limousine' 'small car' 'combi' nan 'coupe' 'suv' 'convertible'
 'other']


In [40]:
# grearbox dictionary
gearbox_dict = {"manuell":"manually", "automatik":"automatic"}

# fuel type dictionary
fueltype_dict = {"lpg":"lpg", "benzin":"petrol", "diesel":"diesel",
                "cng":"cng", "hybrid":"hybrid", "elektro":"electro",
                "andere":"other"}
# unrepaired damage
ud_dict = {"nein":"no", "ja":"yes"}


# mapping gearbox
autos["gearbox"] = autos["gearbox"].map(gearbox_dict)
print(autos["gearbox"].unique())
# mapping fueltype
autos["fueltype"] = autos["fueltype"].map(fueltype_dict)
print(autos["fueltype"].unique())
# mapping unrepaired damage
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(ud_dict)
print(autos["unrepaired_damage"].unique())

[nan]
['lpg' nan 'diesel' 'cng' 'hybrid']
['no' nan 'yes']


In [34]:
# extract datecrawled strings
date = autos["datecrawled"].str[:10]
# remove "-" like "20160321" and return to column "datecrawled"
autos["datecrawled"] = date.str.replace("-","")

# print(autos["datecrawled"]) to check


In [35]:
# Find the most common brand/model combinations
model = autos["model"].value_counts(normalize=True)
print(model.head(5))
# Top 5 model on the eBay is: Gold, Andere, 3er, polo and corsa

def find_brand(model_name):
    print(autos[autos["model"]==model_name].head(1))

find_brand("corsa")



golf      0.083326
andere    0.075818
3er       0.058780
polo      0.036167
corsa     0.035785
Name: model, dtype: float64
   datecrawled                  name  price   abtest vehicletype  \
32    20160320  Corsa_mit_TÜV_5.2016    350  control   small car   

    registration_year   gearbox  powerps  model  odometer_km  \
32               1999  manually        0  corsa       150000   

    registration_month fueltype brand unrepaired_damage           ad_created  \
32                   7   petrol  opel               NaN  2016-03-20 00:00:00   

    postalcode             lastseen  
32       27619  2016-04-06 03:15:20  


#### Top 5 brand/model on the eBay is:
1. Volkswagen/Golf (8.3%)
2. Peugeot/Andere (7.5%)
3. BMW/3er (5.8%)
4. Volkswagen/Polo (3.6%)
5. Opel/Corsa (3.5%)

In [36]:
# spilt the odometer_km into groups
# use aggregation to see if average prices
# follows any patterns based on the milage

# find out frequency of milage and build a list for later use
milage_percentage = autos["odometer_km"].value_counts(normalize=True)
milage_list = milage_percentage.index[:]


# build an empty dictionary to find average price by milage
milage_price = {}

# use a for-loop and boolean method to calculate average price
for milage in milage_list:
    bool_3 = (autos["odometer_km"] == milage)
    A = autos["price"][bool_3].mean().round()
    milage_price[milage] = A
    
# print(milage_price) to check

# establish a dataframe for analysis
mp_series = pd.Series(milage_price)
mp = pd.DataFrame(mp_series, columns = ["mean price"])
print(mp)

        mean price
5000        8874.0
10000      20551.0
20000      18448.0
30000      16609.0
40000      15500.0
50000      13812.0
60000      12385.0
70000      10927.0
80000       9722.0
90000       8465.0
100000      8133.0
125000      6214.0
150000      3768.0


Based on dataframe above, we can seperate data into 2 groups.

For milage less than 5000km, we can see the average price is also lower. We have to go deeper to analyze if there's another variable to affect price, brand and model for example.

For another group which milage over 5000km to 150000km, we can see the more milage car is, the more depreciation it is which means the car is sold at cheaper price.

I'm not sure how much is new car sold in Germany. In my country, most of new cars was sold at about $30000~$40000 tariff included. If you'd like to spend less money and have better quality, you could try to find cars which milage is in about 20000km.

In [44]:
# How much cheaper are cars with damage
# than their non-damage counterparts
# create a percentage 
damage_or_not = autos["unrepaired_damage"].value_counts(normalize=True)
yes_or_no = damage_or_not.index[:]

# create an empty dictionary
damage_price = {}

# use a for-loop and boolean method to calculate average price
for row in yes_or_no:
    bool_4 = (autos["unrepaired_damage"] == row)
    damage_price_list = autos[bool_4]["price"].mean().round()
    damage_price[row] = damage_price_list

#  establish a dataframe for analysis
dp_series = pd.Series(damage_price)
dp = pd.DataFrame(dp_series, columns=["mean price"])

print(dp)


     mean price
no       7164.0
yes      2241.0


In [45]:
7164-2241

4923

Obviously, the car without damage can be sold at better price.

Based on column above, average price of the damaged car is approximately $4900 cheaper than that of the undamaged car.